### PR TITLE
[consensus] Extract block validation from ProposerElection trait

### LINKED ISSUE #1416: [consensus] Extract block validation from ProposerElection trait
ProposerElection main responsibility is the method with which next proposer/leader is selected. We should be able to have multiple implementations, without code duplication. Currently is_valid_proposal doesn't have the same meaning across all implementations.

is_valid_proposal checks that author doesn't create multiple blocks is independent of selection criteria and should be moved out, and implemented once.

### PR DESCRIPTION
**Pull request by ikabiljo**
### Description

Moved is_valid_proposal out of ProposerElection into new BlockAwareProposerElection (suggestions for better name appreciated). This makes all implementations for ProposerElection have the same meaning for is_valid_proposal, which wasn't the case before.

With this, we keep only actual selection of the leader in the ProposerElection trait. That means that tests for those implementations don't need to be bothered by block validation, but only by leader selection (while newly added tests checks for proper block validation)
This reduces duplication in tests (and will reduce merge conclicts / work needed when changing things related to blocks - like this diff that just had to go through all the tests here - https://github.com/aptos-labs/aptos-core/commit/94db7a5702ca8376b759b1771dac1dcc0b0ab7ae )

Closes #1416

### Test Plan
Run all unit tests in liveness folder, included the newly introduced one

### CODE REVIEW

**Comment by ikabiljo** [comment]
for the failing round_manager::round_manager_test::no_vote_on_old_proposal test:

currently, we have 3 implementations of ProposerElection.
Only ReputationElection checked for duplicate blocks, other two don't. I tried to clean that up here, but we have a test - round_manager::round_manager_test::no_vote_on_old_proposal, using RotatingProposer, and actually relying on is_valid_proposal not rejecting duplicate blocks from the same author.

I am not exactly sure what those tests should test, but should rounds on the blocks be different here - https://github.com/aptos-labs/aptos-core/blame/94db7a5702ca8376b759b1771dac1dcc0b0ab7ae/consensus/src/round_manager_test.rs#L375 ?

File: consensus/src/liveness/block_aware_proposer_election.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    // Return if a given proposed block is valid:
+    // - if a given author is a valid candidate for being a proposer
+    // - if this is the first block proposer has submitted in this round
+    // - if it is not old proposal
+    pub fn is_valid_proposal(&self, block: &Block) -> bool {
+        block.author().map_or(false, |author| {
+            let valid_author = self.is_valid_proposer(author, block.round());
+            if !valid_author {
+                println!("Not a valid author");
```
Comment (zekun000) [inline consensus/src/liveness/block_aware_proposer_election.rs:47]: we probably should use info! or debug! with more information, like the round number etc

**Review by zekun000 — APPROVED** [review]

File: consensus/src/liveness/unequivocal_proposer_election.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                error!(
+                    SecurityEvent::InvalidConsensusProposal,
+                    "Proposal is not from valid author {}, expected {} for round {} and id {}",
+                    author,
+                    self.get_valid_proposer(block.round()),
+                    block.round(),
+                    block.id()
+                );
+
+                println!("Not a valid author");
```
Comment (zekun000) [inline consensus/src/liveness/unequivocal_proposer_election.rs:56]: we should remove this print

File: consensus/src/liveness/unequivocal_proposer_election.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    // Return if a given proposed block is valid:
+    // - if a given author is a valid candidate for being a proposer
+    // - if this is the first block proposer has submitted in this round
+    // - if it is not old proposal
+    pub fn is_valid_proposal(&self, block: &Block) -> bool {
+        block.author().map_or(false, |author| {
+            let valid_author = self.is_valid_proposer(author, block.round());
+            if !valid_author {
+                error!(
+                    SecurityEvent::InvalidConsensusProposal,
```
Comment (zekun000) [inline consensus/src/liveness/unequivocal_proposer_election.rs:48]: this is not a security event though, it's possible that just this node has a different committed state so it derives another leader for the round

**Pull request merged**